### PR TITLE
[api] merge interpretation structs.

### PR DESCRIPTION
**Pull request by ikaven1024**
Signed-off-by: yingjinhui <yingjinhui@didiglobal.com>

**What type of PR is this?**

/kind api-change

**What this PR does / why we need it**:
xxxInterpretation structs can be merged.

**Which issue(s) this PR fixes**:
Part of #2371

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/apis/config/v1alpha1/resourceinterpretercustomization_types.go [hunk 1]
```diff
@@ -54,8 +54,28 @@ type CustomizationRules struct {
...
+	//                     dependency = {}
+	//                     dependency.apiVersion = "v1"
+	//                     dependency.kind = "ServiceAccount"
+	//                     dependency.name = desiredObj.spec.serviceAccountName
+	//                     dependency.namespace = desiredObj.namespace
+	//                     dependencies[0] = {}
+	//                     dependencies[0] = dependency
+	//                 end
+	//                 return dependencies
+	//             end
```
Comment (XiShanYongYe-Chang) [inline pkg/apis/config/v1alpha1/resourceinterpretercustomization_types.go:74]: This comment seems not to match the current operation.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/assign @RainbowMango

File: pkg/apis/config/v1alpha1/resourceinterpretercustomization_types.go [hunk 2]
```diff
@@ -54,8 +54,28 @@ type CustomizationRules struct {
...
+	// Interpretation.LuaScript only holds the function body part, take the `desiredObj` and `observedObj` as the function
+	// parameters or global variables, and finally returns a retained specification.
 	// +optional
-	Retention *LocalValueRetention `json:"retention,omitempty"`
+	Retention *Interpretation `json:"retention,omitempty"`
```
Comment (chaunceyjiang) [inline pkg/apis/config/v1alpha1/resourceinterpretercustomization_types.go:78]: Will the degree of coupling be too high?

**Comment by RainbowMango** [comment]
> xxxInterpretation structs can be merged.

Why? What's the benefit?

**Comment by ikaven1024** [comment]
> > xxxInterpretation structs can be merged.
> 
> Why? What's the benefit?

I can handle the script in one function.

https://github.com/ikaven1024/karmada/blob/609351cf6281e7da51f6f8e608b2a29d8776f47a/pkg/resourceinterpreter/configurableinterpreter/configurable.go#L154-L190

**Comment by RainbowMango** [comment]
The reason why I didn't use a common struct to describe the rules is `script` isn't the only way to represent the rules.
Each interpreter operation may be described in another easier way instead of script.  The idea didn't present at current APIs, but you can see it in the [original proposal](https://github.com/karmada-io/karmada/tree/master/docs/proposals/configurable-local-value-retention#new-config-api). Except the lua script, people can specify the rule by `Json` rule:
```
// Fields indicates the fields that should be retained.
	// Each field describes a field in JsonPath format.
	// +optional
	Fields []string `json:"fields,omitempty"`

	// RetentionLua holds the Lua script that is used to retain runtime values to the desired specification.
	// The script should implement a function just like:
	//
	// function Retain(desiredObj, runtimeObj)
	//     desiredObj.spec.fieldFoo = runtimeObj.spec.fieldFoo
	//     return desiredObj
	// end
	//
	// RetentionLua only holds the function body part, take the `desiredObj` and `runtimeObj` as the function parameters or
	// global parameters, and finally retains a retained specification.
	// +optional
	RetentionLua string `json:"retentionLua,omitempty"`
```

In other words, each interpreter operation may have different customization methods respectively. I'm afraid using a common struct would be very hard to do so.

**Comment by ikaven1024** [comment]
>

**Comment by ikaven1024** [comment]
> The reason why I didn't use a common struct to describe the rules is `script` isn't the only way to represent the rules. Each interpreter operation may be described in another easier way instead of script. The idea didn't present at current APIs, but you can see it in the [original proposal](https://github.com/karmada-io/karmada/tree/master/docs/proposals/configurable-local-value-retention#new-config-api). Except the lua script, people can specify the rule by `Json` rule:
> 
> ```
> // Fields indicates the fields that should be retained.
> 	// Each field describes a field in JsonPath format.
> 	// +optional
> 	Fields []string `json:"fields,omitempty"`
> 
> 	// RetentionLua holds the Lua script that is used to retain runtime values to the desired specification.
> 	// The script should implement a function just like:
> 	//
> 	// function Retain(desiredObj, runtimeObj)
> 	//     desiredObj.spec.fieldFoo = runtimeObj.spec.fieldFoo
> 	//     return desiredObj
> 	// end
> 	//
> 	// RetentionLua only holds the function body part, take the `desiredObj` and `runtimeObj` as the function parameters or
> 	// global parameters, and finally retains a retained specification.
> 	// +optional
> 	RetentionLua string `json:"retentionLua,omitempty"`
> ```
> 
> In other words, each interpreter operation may have different customization methods respectively. I'm afraid using a common struct would be very hard to do so.

OK, Keep the api.
